psrqpy/search.py

```diff
@@ -18,586 +18,2043 @@
+        self.update(XX, name='XX')
+        self.update(YY, name='YY')
+        self.update(ZZ, name='ZZ')
+
+        # set DMSINB
+        if 'DM' in self.columns:
+            DM = self.catalogue['DM']
+            GB = self.catalogue['GB']
+
+            DMSINB = np.full(self.catalogue_len, np.nan)
+
+            idx = np.isfinite(GB) & np.isfinite(DM)
+            DMSINB[idx] = DM[idx]*np.sin(np.deg2rad(GB[idx]))
+
+            self.update(DMSINB, name='DMSINB')
+
+        # galactic proper motion
+        if not np.all([p in self.columns for p in ['PMB', 'PML']]):
+            reqpar = ['PMRA', 'PMDEC']
+            if np.all([p in self.columns for p in reqpar]):
+                PMRA = self.catalogue['PMRA'].values.copy()
+                PMDEC = self.catalogue['PMDEC'].values.copy()
+
+                PMB = np.full(self.catalogue_len, np.nan)
+                PML = np.full(self.catalogue_len, np.nan)
+
+                idx = (np.isfinite(PMRA) & np.isfinite(PMDEC) &
+                       np.isfinite(RAJD) & np.isfinite(DECJD) &
+                       np.isfinite(DIST))
+
+                sc = ICRS(RAJD[idx]*aunits.deg, DECJD[idx]*aunits.deg,
+                          distance=DIST[idx]*aunits.kpc,
+                          pm_ra_cosdec=PMRA[idx]*aunits.mas/aunits.yr,
+                          pm_dec=PMDEC[idx]*aunits.mas/aunits.yr
+                          ).transform_to(Galactic())
+
+                PMB[idx] = sc.pm_b.value
+                PML[idx] = sc.pm_l_cosb.value
+
+                self.update(PMB, name='PMB')
+                self.update(PML, name='PML')
+
+    def derived_binary(self):
+        """
+        Calculate derived binary system parameters.
+        """
+
+        MASS_PSR = 1.35  # canonical pulsar mass (solar masses)
 
-        if loadfromfile:
-            self.load(loadfromfile)
+        # derive mass function
+        reqpars = ['A1', 'PB']
+        if np.all([p in self.columns for p in reqpars]):
+            A1 = self.catalogue['A1'].values.copy()*c.value
+            PB = self.catalogue['PB'].values.copy()*86400.
+
+            idx = np.isfinite(A1) & np.isfinite(PB)
+
+            MASSFN = np.full(self.catalogue_len, np.nan)
+            MASSFN[idx] = (4.*np.pi**2/GM_sun.value)*A1[idx]**3/(PB[idx]**2)
+
+            self.update(MASSFN, name='MASSFN')
+
+            # derive minimum, median and 90% UL for mass
+            MINMASS = np.full(self.catalogue_len, np.nan)
+            MEDMASS = np.full(self.catalogue_len, np.nan)
+            UPRMASS = np.full(self.catalogue_len, np.nan)
+            from scipy.optimize import newton
+
+            def solfunc(m2, sini, mf, m1):
+                return (m1 + m2)**2 - (m2*sini)**3/mf
+
+            SINI_MIN = 1.0  # inclination for minimum mass
+            SINI_MED = 0.866025403  # inclination of 60deg for median mass
+            SINI_90 = 0.438371146   # inclination for 90% UL mass
+            for i, mf in enumerate(MASSFN):
```

Comment:
I've looked at a couple of examples and they were consistent with what is in the ATNF and produced by `psrcat`, although in general the masses all looked rather low. It may be that what's currently done in `psrcat`, which I've copied, needs to be updated. So, we could keep the values we currently have for consistency, but if there's another way of doing things we could also add that into `psrqpy`.

Reply:
I might take a more careful look there, and see what's done in the literature for better reference.

Reply:
So I think I was misunderstanding what `newton` actually does. It doesn't allow `m_p` to vary, correct? Just `m_c`? That should all be fine, and I haven't found anything different in the literature.